### PR TITLE
Refactor States To Allow for Single Cached Hasher

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Refactor

**What does this PR do? Why is it needed?**

- [x] Currently our cached state hasher is duplicated across 3 states. This PR instantiates it in the `stateutil` package so that
only a single implementation of it exists. This reduces duplication as now it only has to be implemented once rather than multiple times for each state type.

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

**Review by rkapka — APPROVED** [review]

**Review by potuz — COMMENTED** [review]
I went though it and didn't find anything, but seems hard to review this seriously without spending the whole day. If the Spec tests are comprehensive for all states, and this passes spec tests I'd be inclined to approve this.

File: beacon-chain/state/stateutil/state_hasher.go [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/crypto/hash"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/encoding/ssz"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+	"go.opencensus.io/trace"
+)
+
+var (
+	leavesCache = make(map[string][][32]byte, params.BeaconConfig().BeaconStateMergeFieldCount)
```
Comment (terencechain) [inline beacon-chain/state/stateutil/state_hasher.go:19]: Are we just always taking the highest field count here? if yes, could we please add a comment to clarify?

File: beacon-chain/state/stateutil/field_root_attestation.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/prysmaticlabs/prysm/config/features"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/crypto/hash"
+	"github.com/prysmaticlabs/prysm/encoding/ssz"
+	ethpb "github.com/prysmaticlabs/prysm/proto/prysm/v1alpha1"
+)
+
+// RootsArrayHashTreeRoot computes the Merkle root of arrays of 32-byte hashes, such as [64][32]byte
+// according to the Simple Serialize specification of Ethereum.
+func RootsArrayHashTreeRoot(vals [][]byte, length uint64, fieldName string) ([32]byte, error) {
```
Comment (terencechain) [inline beacon-chain/state/stateutil/field_root_attestation.go:18]: This is only used in tests, can we delete this function?

File: beacon-chain/state/stateutil/field_root_attestation.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	res := ssz.MixInLength(attsRootsRoot, attsLenRoot)
+	return res, nil
+}
+
+func (h *stateRootHasher) pendingAttestationRoot(hasher ssz.HashFn, att *ethpb.PendingAttestation) ([32]byte, error) {
+	if att == nil {
+		return [32]byte{}, errors.New("nil pending attestation")
+	}
+	// Marshal attestation to determine if it exists in the cache.
+	enc := PendingAttEncKey(att)
```
Comment (terencechain) [inline beacon-chain/state/stateutil/field_root_attestation.go:66]: `PendingAttEncKey` no longer needs to be exported

File: beacon-chain/state/stateutil/field_root_attestation.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+	// Check if it exists in cache:
+	if h.rootsCache != nil {
+		if found, ok := h.rootsCache.Get(string(enc)); found != nil && ok {
+			return found.([32]byte), nil
+		}
+	}
+
+	res, err := PendingAttRootWithHasher(hasher, att)
```
Comment (terencechain) [inline beacon-chain/state/stateutil/field_root_attestation.go:75]: `PendingAttRootWithHasher` no longer needs to be exported

File: beacon-chain/state/stateutil/field_root_eth1.go [hunk 5]
```diff
@@ -1,35 +1,34 @@
...
 // a BeaconBlockHeader struct according to the eth2
 // Simple Serialize specification.
-func eth1Root(hasher ssz.HashFn, eth1Data *ethpb.Eth1Data) ([32]byte, error) {
+func Eth1Root(hasher ssz.HashFn, eth1Data *ethpb.Eth1Data) ([32]byte, error) {
 	if eth1Data == nil {
 		return [32]byte{}, errors.New("nil eth1 data")
 	}
 
-	enc := stateutil.Eth1DataEncKey(eth1Data)
+	enc := Eth1DataEncKey(eth1Data)
```
Comment (terencechain) [inline beacon-chain/state/stateutil/field_root_eth1.go:18]: `Eth1DataEncKey` no longer needs to be exported

File: beacon-chain/state/stateutil/field_root_validator.go [hunk 6]
```diff
@@ -59,15 +68,15 @@ func (h *stateRootHasher) validatorRoot(hasher ssz.HashFn, validator *ethpb.Vali
 		return [32]byte{}, errors.New("nil validator")
 	}
 
-	enc := stateutil.ValidatorEncKey(validator)
+	enc := ValidatorEncKey(validator)
```
Comment (terencechain) [inline beacon-chain/state/stateutil/field_root_validator.go:71]: `ValidatorEncKey` no longer needs to be exported

File: beacon-chain/state/stateutil/state_hasher.go [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+	lock        sync.RWMutex
+)
+
+const cacheSize = 100000
+
+// NocachedHasher references a hasher that will not utilize a cache.
+var NocachedHasher *stateRootHasher
+
+// CachedHasher references a hasher that will utilize a roots cache.
+var CachedHasher *stateRootHasher
```
Comment (terencechain) [inline beacon-chain/state/stateutil/state_hasher.go:30]: One minor downside with this is we have to export these hashers to be shared with state pkgs

**Review by terencechain — COMMENTED** [review]
Nice work!

File: beacon-chain/state/stateutil/field_root_attestation.go (hunk 2 above)
Comment (nisdas) [reply beacon-chain/state/stateutil/field_root_attestation.go:18]: This is used in runtime as part of `computeFieldRoots`. This is why we continue using it, this should be removed later but I will need a bit more data to be sure on this.

**Review by terencechain — DISMISSED** [review]
🎉

File: beacon-chain/state/stateutil/state_hasher.go [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+
+// stateRootHasher defines an object through which we can
+// hash the different fields in the state with a few cached layers.
+type stateRootHasher struct {
+	rootsCache *ristretto.Cache
+}
+
+// ComputeFieldRootsWithHasherPhase0 hashes the provided phase 0 state and returns its respective field roots.
+func (h *stateRootHasher) ComputeFieldRootsWithHasherPhase0(ctx context.Context, state *ethpb.BeaconState) ([][]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherPhase0")
```
Comment (terencechain) [inline beacon-chain/state/stateutil/state_hasher.go:56]:
```suggestion
	_, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherPhase0")
```

deep source complains

File: beacon-chain/state/stateutil/state_hasher.go [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "could not compute finalized checkpoint merkleization")
+	}
+	fieldRoots[20] = finalRoot[:]
+	return fieldRoots, nil
+}
+
+// ComputeFieldRootsWithHasherAltair hashes the provided altair state and returns its respective field roots.
+func (h *stateRootHasher) ComputeFieldRootsWithHasherAltair(ctx context.Context, state *ethpb.BeaconStateAltair) ([][]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherAltair")
```
Comment (terencechain) [inline beacon-chain/state/stateutil/state_hasher.go:204]:
```suggestion
	_, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherAltair")
```

File: beacon-chain/state/stateutil/state_hasher.go [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+		return nil, errors.Wrap(err, "could not compute sync committee merkleization")
+	}
+	fieldRoots[23] = nextSyncCommitteeRoot[:]
+
+	return fieldRoots, nil
+}
+
+// ComputeFieldRootsWithHasherMerge hashes the provided merge state and returns its respective field roots.
+func (h *stateRootHasher) ComputeFieldRootsWithHasherMerge(ctx context.Context, state *ethpb.BeaconStateMerge) ([][]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherMerge")
```
Comment (terencechain) [inline beacon-chain/state/stateutil/state_hasher.go:374]:
```suggestion
	_, span := trace.StartSpan(ctx, "hasher.ComputeFieldRootsWithHasherMerge")
```

**Review by terencechain — APPROVED** [review]

**Pull request merged**